Documentation/ProvenHeaders.md

```diff
@@ -0,0 +1,229 @@
+# ProvenHeaders specifications
+## Table of contents
+
+- [Problem overview](#problem-overview)
+
+- [Solution abstract overview](#solution-abstract-overview)
+
+- [Basic constructs](#basic-constructs)
+
+- [Structure of proven header](#structure-of-proven-header)
+
+- [Assumptions](#assumptions)
+
+- [Best chain selection](#best-chain-selection)
+
+- [Changes required for the C# node](#changes-required-for-the-c-node)
+
+  ​
+
+
+## Problem overview
+
+Syncing with headers is used for quick blockchain download. Node can ask its peers for headers using *getheaders* message and asked peer will send up to 2000 headers per *headers* message. When the headers are received they are used in order to ask peers for blocks that are represented by those headers. This protocol is new and used on bitcoin network because it has some advantages in terms of resources usage over using *inv* messages.
+
+On PoW chains headers can't be easily faked because creating a valid header requires spending a lot of hashing power to meet the PoW difficulty target.
+
+On PoS networks like Stratis it is easy to construct a fake chain of headers of any length and almost arbitrary chainwork. 
+
+With our C# node implementation, if an attacker constructs a fake chain of headers that has more work than the valid chain and sends it to a node that node will switch to a fake chain of headers. After switching the node will ask for blocks that are represented by the fake headers but those blocks will never be received so the node will stay out of sync with the network until the moment network eventually produces a longer chain and the node switches back.
+
+*Therefore it is possible to perform an attack on any node that uses getheaders and keep it out of sync for a long period of time.* 
+
+## Solution abstract overview
+
+In order to prevent attacks using fake headers we propose construction of proven headers. 
+
+New type of network messages needs to be introduced to allow syncing using proven headers -
+
+1.  *provhdr* - `ProvenHeadersPayload` which contains list of up to 2000 proven headers.
+2. *sendprovhdr* - `SendProvenHeadersPayload` which informs the peer that we are only willing to sync using the proven headers and not the old type of headers.
+3. *getprovhdr* - `GetProvenHeadersPayload`  which requests proven headers using a similar mechanism as the `getheaders` protocol message.
+
+
+
+
+
+## Basic constructs
+
+#### Coinstake age
+
+ `coinstakeAge` is a consensus constant that specifies how old (confirmation-wise) a UTXO has to be in order to be allowed to be used as a coinstake's kernel. 
+
+This means that if we are synced till the block of height `x` and we receive a proven header that represents a block within the range from `x + 1` to `x + coinstakeAge` then we have enough data to validate the coinstake's kernel because a valid kernel couldn't have been originated in a block with height greater than `coinstakeAge`. 
```

Comment:
```suggestion
This means that if we are synced up to the block of height `x` and we receive a proven header that represents a block within the range from `x + 1` to `x + coinstakeAge` then we have enough data to validate the coinstake's kernel because a valid kernel couldn't have been originated in a block with height greater than `coinstakeAge`. 
```